packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol

```diff
@@ -258,101 +286,87 @@ contract OptimisticGovernor is Module, Lockable {
+        require(assertionId != bytes32(0), "Proposal hash does not exist");
+
+        // Detect Optimistic Asserter upgrade by checking if it has the matching assertionId.
+        require(optimisticAsserter.getAssertion(assertionId).asserter == address(0), "OA upgrade not detected");
 
+        // Remove proposal hash and assertionId so that transactions can be re-proposed if needed.
         delete proposalHashes[_proposalHash];
-        emit ProposalDeleted(_proposalHash, msg.sender, "DeletedByOwner");
+        delete assertionIds[assertionId];
+
+        emit ProposalDeleted(_proposalHash, assertionId);
     }
 
     /**
-     * @notice Method to allow anyone to delete a proposal that was rejected.
-     * @param _proposalHash the hash of the proposal being deleted.
+     * @notice Callback to automatically delete a proposal that was disputed.
+     * @param assertionId the identifier of the disputed assertion.
      */
-    function deleteRejectedProposal(bytes32 _proposalHash) external {
-        // Check that proposal exists and was not already deleted.
-        require(proposalHashes[_proposalHash] != 0, "Proposal does not exist");
-
-        // Add the proposal hash to ancillary data.
-        bytes memory ancillaryData = AncillaryData.appendKeyValueBytes32("", PROPOSAL_HASH_KEY, _proposalHash);
-
-        // Get the original proposal time.
-        uint256 _originalTime = proposalHashes[_proposalHash];
-
-        // This will revert if the price has not been settled and cannot currently be settled.
-        int256 price = optimisticOracle.settleAndGetPrice(identifier, _originalTime, ancillaryData);
-
-        // Check that proposal was rejected.
-        require(price != PROPOSAL_VALID_RESPONSE, "Proposal was not rejected");
-
-        // Delete the proposal.
-        delete proposalHashes[_proposalHash];
-        emit ProposalDeleted(_proposalHash, msg.sender, "Rejected");
+    function assertionDisputedCallback(bytes32 assertionId) external {
+        // In order to optimize for happy path, the assertionId is validated for potential spoofing only in the
+        // deleteProposalOnUpgrade call. Genuine Optimistic Asserter should always pass a valid assertionId that has a
+        // matching proposalHash in this contract.
+        bytes32 proposalHash = assertionIds[assertionId];
+
+        // Callback should only be called by the Optimistic Asserter. Address would not match in case of contract
+        // upgrade, thus try deleting the proposal through deleteProposalOnUpgrade function that should revert if
+        // address mismatch was not caused by an Optimistic Asserter upgrade.
+        if (msg.sender == address(optimisticAsserter)) {
+            // Delete the disputed proposal and associated assertionId.
+            delete proposalHashes[proposalHash];
+            delete assertionIds[assertionId];
+
+            emit ProposalDeleted(proposalHash, assertionId);
+        } else deleteProposalOnUpgrade(proposalHash);
```

Comment:
does this other flow not need any permissioning or entry checks on it?

Reply:
if `assertionId` was invalid by somebody spoofing this callback then `proposalHash` would be zero and that would revert within `deleteProposalOnUpgrade`.

While if somebody was spoofing callback with valid `assertionId` then `deleteProposalOnUpgrade` would also revert due to OA having a matching assertion with non-zero asserter.